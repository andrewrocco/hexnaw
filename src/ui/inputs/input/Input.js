import React from 'react';
import { Icon } from 'ui/icons';

import {
  StyledClose,
  StyledInput,
  StyledLabel,
  StyledWrap,
} from './Input.styles';

export const getValidHex = value => (
  (/^[a-fA-F0-9]{6}$|^[a-fA-F0-9]{3}$/).test(value)
    ? `#${value}`
    : 'transparent'
);

export const Input = ({ field, onRemove, ...props }) => (
  <StyledWrap hexValue={getValidHex(field.value)}>
    <StyledInput
      autoComplete="off"
      autoCapitalize="off"
      spellCheck="false"
      type="text"
      {...field}
      {...props}
    />
    <StyledLabel htmlFor={field.name}>#</StyledLabel>
    {onRemove && (
      <StyledClose type="button" onClick={onRemove}>
        <Icon icon="Close" />
      </StyledClose>
    )}
  </StyledWrap>
);
