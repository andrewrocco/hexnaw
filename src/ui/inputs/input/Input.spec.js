import React from 'react';
import { Input, getValidHex } from './Input';
import 'jest-styled-components';

describe('Input', () => {
  const onRemove = jest.fn();
  const subject = mount(
    <Input
      field={{ value: 'fff' }}
      onChange={() => {}}
      onRemove={onRemove}
    />
  );

  context('when nothing specified', () => {
    it('should render correctly', () => {
      expect(subject).toMatchSnapshot();
    });
  });

  context('getValidHex helper', () => {
    it('returns valid hex value with # prefix', () => {
      expect(getValidHex('bbb')).toBe('#bbb');
    });

    it('returns "transparent" if hex value is invalid', () => {
      expect(getValidHex('meh')).toBe('transparent');
    });
  });

  // TODO: re-enable after version bump of jest-styled-components
  // context('when hexValue is passed', () => {
  //   it('renders the hex value in the after pseudo selector', () => {
  //     expect(subject.find('div').first()).toHaveStyleRule('background', '#fff', {
  //       modifier: ':after',
  //     });
  //   });
  // });

  context('when onRemove function is passed', () => {
    it('renders the close button', () => {
      expect(subject.find('button')).toHaveLength(1);
    });

    it('calls onRemove function when close button is clicked', () => {
      subject.find('button').simulate('click');
      expect(onRemove).toHaveBeenCalledTimes(1);
    });
  });
});
