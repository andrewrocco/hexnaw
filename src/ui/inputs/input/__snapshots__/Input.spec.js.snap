// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input when nothing specified should render correctly 1`] = `
<Input
  field={
    Object {
      "value": "fff",
    }
  }
  onChange={[Function]}
  onRemove={[MockFunction]}
>
  <styled.div
    hexValue="#fff"
  >
    <div
      className="sc-19402xi-3 dIarrY"
    >
      <styled.input
        autoCapitalize="off"
        autoComplete="off"
        onChange={[Function]}
        spellCheck="false"
        type="text"
        value="fff"
      >
        <input
          autoCapitalize="off"
          autoComplete="off"
          className="sc-19402xi-1 dcgzjL"
          onChange={[Function]}
          spellCheck="false"
          type="text"
          value="fff"
        />
      </styled.input>
      <styled.label>
        <label
          className="sc-19402xi-2 hSEqrQ"
        >
          #
        </label>
      </styled.label>
      <styled.button
        onClick={[MockFunction]}
        type="button"
      >
        <button
          className="sc-19402xi-0 eBjPIc"
          onClick={[MockFunction]}
          type="button"
        >
          <Icon
            icon="Close"
          >
            <svg
              viewBox="0 0 11 11"
            >
              <path
                d="M5.5 3.732l2.616-2.616c.488-.488 1.28-.488 1.768 0s.488 1.28 0 1.768L7.268 5.5l2.616 2.616c.488.488.488 1.28 0 1.768s-1.28.488-1.768 0L5.5 7.268 2.884 9.884c-.488.488-1.28.488-1.768 0s-.488-1.28 0-1.768L3.732 5.5 1.116 2.884c-.488-.488-.488-1.28 0-1.768s1.28-.488 1.768 0L5.5 3.732z"
              />
            </svg>
          </Icon>
        </button>
      </styled.button>
    </div>
  </styled.div>
</Input>
`;
